File: src/09.tree/03.101.go
package tree

func IsSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return dfs(root.Left, root.Right)
}

func dfs(l *TreeNode, r *TreeNode) bool {
	isL := l == nil
	isR := r == nil

	if isL && isR {
		return true
	}

	if isL || isR {
		return false
	}

	if l.Val != r.Val {
		return false
	}

	return dfs(l.Left, r.Right) && dfs(l.Right, r.Left)
}

File: src/09.tree/03.101.ts
import { TreeNode } from "./01.100";

export function isSymmetric(root: TreeNode | null): boolean {
    if (!root) {
        return true;
    }
    return dfs(root.left, root.right);
}

function dfs(l: TreeNode | null, r: TreeNode | null): boolean {
    if (!l && !r) {
        return true;
    }
    if (!l || !r) {
        return false;
    }
    if (l.val !== r.val) {
        return false;
    }

    return dfs(l.left, r.right) && dfs(l.right, r.left);
}
